qos/cosmos/request_validator_rest.go

```diff
@@ -0,0 +1,246 @@
+		protocolErrorResponseBuilder: buildRESTProtocolErrorResponse(),
+		// Protocol-level request error observation is the same for JSONRPC and REST.
+		protocolErrorObservationBuilder: buildProtocolErrorObservation,
+	}, true
+}
+
+func buildRESTServicePayload(
+	rpcType sharedtypes.RPCType,
+	httpRequestURL *url.URL,
+	httpRequestMethod string,
+	httpRequestBody []byte,
+) protocol.Payload {
+	path := httpRequestURL.Path
+	if httpRequestURL.RawQuery != "" {
+		path += "?" + httpRequestURL.RawQuery
+	}
+
+	return protocol.Payload{
+		Data:            string(httpRequestBody),
+		Method:          httpRequestMethod,
+		TimeoutMillisec: defaultRESTRequestTimeoutMillisec,
+		// Add the RPCType hint, so protocol sets correct HTTP headers for the endpoint.
+		RPCType: rpcType,
+		// Set the request path, including raw query, if used.
+		Path: path,
+	}
+}
+
+func getRESTRequestEndpointResponseValidator(requestPath string) func(polylog.Logger, []byte) response {
+	// Delegate the unmarshaling/validation of endpoint response to the specialized REST unmarshaler.
+	return func(logger polylog.Logger, endpointResponseBz []byte) response {
+		return unmarshalRESTRequestEndpointResponse(logger, requestPath, endpointResponseBz)
+	}
+}
+
+func (rv *requestValidator) buildRESTRequestObservations(
+	rpcType sharedtypes.RPCType,
+	httpRequestURL *url.URL,
+	httpRequestMethod string,
+	httpRequestBody []byte,
+	servicePayload protocol.Payload,
+) *qosobservations.CosmosRequestObservations {
+
+	// Determine content type from headers if available, otherwise empty
+	contentType := ""
+	// Note: We don't have access to headers here, but this would be where we'd extract it
+
+	return &qosobservations.CosmosRequestObservations{
+		ChainId:       rv.chainID,
+		ServiceId:     string(rv.serviceID),
+		RequestOrigin: qosobservations.RequestOrigin_REQUEST_ORIGIN_ORGANIC,
+		RequestProfile: &qosobservations.CosmosRequestProfile{
+			BackendServiceDetails: &qosobservations.BackendServiceDetails{
+				BackendServiceType: convertToProtoBackendServiceType(rpcType),
+				SelectionReason:    "REST path detection",
+			},
+			ParsedRequest: &qosobservations.CosmosRequestProfile_RestRequest{
+				RestRequest: &qosobservations.RESTRequest{
+					ApiPath:       httpRequestURL.Path,
+					HttpMethod:    httpRequestMethod,
+					ContentType:   contentType,
+					PayloadLength: uint32(len(httpRequestBody)),
+				},
+			},
+		},
+	}
+}
+
+// TODO_TECHDEBT(@adshmh): Review the expected user experience on protocol errors in REST requests.
+func buildRESTProtocolErrorResponse() func(logger polylog.Logger) gateway.HTTPResponse {
+	return func(logger polylog.Logger) gateway.HTTPResponse {
+		// For REST requests, we return a JSON-RPC error response with null ID
+		// TODO_TECHDEBT(@adshmh): Consider returning proper REST error response format
+		errorResp := jsonrpc.NewErrResponseInternalErr(
+			jsonrpc.ID{}, // use null as ID.
```

Comment:
Ditto here. Perhaps we should determine the error response based on RPC type since CometBFT always returns JSON-RPC responses even for REST-like requests.

For CosmosSDK I agree that perhaps it should be REST error response however.